File: NESDeltaCore/Bridge/NESEmulatorBridge.m
//
//  NESEmulatorBridge.m
//  NESDeltaCore
//
//  Created by Riley Testut on 2/25/18.
//  Copyright © 2018 Riley Testut. All rights reserved.
//

#import "NESEmulatorBridge.h"

@interface NESEmulatorBridge ()

@property (nonatomic, copy, nullable, readwrite) NSURL *gameURL;

@end

@implementation NESEmulatorBridge
@synthesize audioRenderer = _audioRenderer;
@synthesize videoRenderer = _videoRenderer;
@synthesize saveUpdateHandler = _saveUpdateHandler;

+ (instancetype)sharedBridge
{
    static NESEmulatorBridge *_emulatorBridge = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _emulatorBridge = [[self alloc] init];
    });
    
    return _emulatorBridge;
}

#pragma mark - Emulation State -

- (void)startWithGameURL:(NSURL *)gameURL
{
}

- (void)stop
{
}

- (void)pause
{
    
}

- (void)resume
{
    
}

#pragma mark - Game Loop -

- (void)runFrame
{
}

#pragma mark - Inputs -

- (void)activateInput:(NSInteger)input
{
    
}

- (void)deactivateInput:(NSInteger)input
{
    
}

- (void)resetInputs
{
    
}

#pragma mark - Save States -

- (void)saveSaveStateToURL:(NSURL *)url
{
    
}

- (void)loadSaveStateFromURL:(NSURL *)url
{
    
}

#pragma mark - Game Saves -

- (void)saveGameSaveToURL:(NSURL *)url
{
    
}

- (void)loadGameSaveFromURL:(NSURL *)url
{
    
}

#pragma mark - Cheats -

- (BOOL)addCheatCode:(NSString *)cheatCode type:(NSString *)type
{
    return YES;
}

- (void)resetCheats
{
    
}

- (void)updateCheats
{
    
}

@end

File: NESDeltaCore/Bridge/NESEmulatorBridge.swift
//
//  NESEmulatorBridge.swift
//  NESDeltaCore
//
//  Created by Riley Testut on 6/1/18.
//  Copyright © 2018 Riley Testut. All rights reserved.
//

import Foundation
import DeltaCore

#if NATIVE
import NESBridge
#endif

class NESEmulatorBridge: AdaptableDeltaBridge
{
    public static let shared = NESEmulatorBridge()
    
    private var isDatabasePrepared = false
    
    override var adapter: EmulatorBridging {
#if !NATIVE
        let scriptURL = Bundle.module.url(forResource: "nestopia", withExtension: "html")!
        
        let adapter = JSCoreAdapter(prefix: "NES", fileURL: scriptURL)
        adapter.emulatorCore = self.emulatorCore
        return adapter
#else
        return NativeCoreAdapter(
            frameDuration: NESFrameDuration,
            start: NESStartEmulation,
            stop: NESStopEmulation,
            pause: {},
            resume: {},
            runFrame: NESRunFrame,
            activateInput: NESActivateInput,
            deactivateInput: NESDeactivateInput,
            resetInputs: NESResetInputs,
            saveSaveState: NESSaveSaveState,
            loadSaveState: NESLoadSaveState,
            saveGameSave: NESSaveGameSave,
            loadGameSave: NESLoadGameSave,
            addCheatCode: NESAddCheatCode,
            resetCheats: NESResetCheats,
            updateCheats: {},
            setAudioCallback: NESSetAudioCallback,
            setVideoCallback: NESSetVideoCallback,
            setSaveCallback: NESSetSaveCallback)
#endif
    }
    
    override func start(withGameURL gameURL: URL)
    {
#if NATIVE
        if !self.isDatabasePrepared
        {
            let databaseURL = Bundle.module.url(forResource: "NstDatabase", withExtension: "xml")!
            databaseURL.withUnsafeFileSystemRepresentation { NESInitialize($0!) }
            
            self.isDatabasePrepared = true
        }
#endif
        
        super.start(withGameURL: gameURL)
    }
}
